src/compiler/binder.ts

```diff
@@ -1138,8 +1138,8 @@ namespace ts {
             }
             else if (node.kind === SyntaxKind.ArrayLiteralExpression) {
                 for (const e of (<ArrayLiteralExpression>node).elements) {
-                    if (e.kind === SyntaxKind.SpreadElementExpression) {
-                        bindAssignmentTargetFlow((<SpreadElementExpression>e).expression);
+                    if (e.kind === SyntaxKind.SpreadExpression) {

```

Comment:
`[ x, ...y]` produces a `SpreadExpression` (now `SpreadElement`) for `...y`. `{ x: 1, ...y }` produces `SpreadElementExpression` (now `SpreadAssignment`) for `...y`. So the nodes are actually different for arrays vs objects.